### PR TITLE
refactor!: extract instrumentation base from api

### PR DESCRIPTION
**Pull request by robertlaurin**
Extracts the instrumentation base and registry modules from the API.   Neither is part of the API spec so we are extracting them prior to the 1.0 release.

### CODE REVIEW

File: instrumentation/base/lib/opentelemetry/instrumentation.rb [hunk 1]
```diff
@@ -4,12 +4,20 @@
...
 module OpenTelemetry
   # The instrumentation module contains functionality to register and install
   # instrumentation
   module Instrumentation
+    extend self
+
+    # @return [Instrumentation::Registry] registry containing all known
+    #  instrumentation
+    def registry
+      @registry ||= Instrumentation::Registry.new
```
Comment (fbogsany) [inline instrumentation/base/lib/opentelemetry/instrumentation.rb:20]:
```suggestion
    # @return [Registry] registry containing all known
    #  instrumentation
    def registry
      @registry ||= Registry.new
```

File: instrumentation/base/README.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
+# OpenTelemetry Base Instrumentation
+
+Todo:
```
Comment (fbogsany) [inline instrumentation/base/README.md:3]: - [ ] Describe how this is to be used.

File: .toys/.data/releases.yml [hunk 3]
```diff
@@ -31,6 +31,10 @@ commit_lint:
 #  *  changelog_path: Path to CHANGLEOG.md relative to the gem directory.
 #     (Required only if it is not in the expected location.)
 gems:
+  - name: opentelemetry-instrumentation-base
+    directory: instrumentation/base
+    version_constant: [OpenTelemetry, Instrumentation, Base, VERSION]
```
Comment (fbogsany) [inline .toys/.data/releases.yml:36]:
```suggestion
    version_constant: [OpenTelemetry, Instrumentation, VERSION]
```

**Review by ahayworth — APPROVED** [review]
I really like the idea of moving this out into its own gem. I personally think `Instrumentation::Base` is a nice thing that library authors _should_ use, but it's nice to move it out on its own - since it's not strictly necessary. 😄

File: instrumentation/base/README.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
+# OpenTelemetry Base Instrumentation
```
Comment (fbogsany) [inline instrumentation/base/README.md:1]:
```suggestion
# OpenTelemetry Instrumentation Base
```

**Review by ericmustin — APPROVED** [review]
I _think_ this lgtm? it's possible i missed something as this has a lot of files changed but overall i think it's the right approach

File: instrumentation/base/README.md [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+```
+gem install opentelemetry-instrumentation-base
+```
+
+Or, if you use [bundler][bundler-home], include `opentelemetry-instrumentation-base` in your `Gemfile`.
+
+### For SDK Authors
+
+The following is a simplified demonstration of how the `OpenTelemetry::Instrumentation.registry` can be used by an SDK to install auto-instrumentation gems as part of it's configuration.  This should not be used as an example of how to implement an SDK as there are large omissions.  For an example of a complete implementation see the `opentelemetry-sdk` gem.
```
Comment (fbogsany) [inline instrumentation/base/README.md:17]:
```suggestion
The following is a simplified demonstration of how the `OpenTelemetry::Instrumentation.registry` can be used by an SDK to install auto-instrumentation gems as part of its configuration.  This should not be used as an example of how to implement an SDK as there are large omissions.  For an example of a complete implementation see the `opentelemetry-sdk` gem.
```

**Review by fbogsany — APPROVED** [review]
Minor typo, otherwise LGTM 👍

**Pull request merged**